### DISCUSSION TITLE
How to nest elements from different modules (for modularization)?

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1040] [Q&A] [answered]
### Question

Hi All,

I am trying to follow the steps from the **Modularization** example which is already listed under the Examples section. However, I think I am unable to follow the logic properly and hence not being able to re-create the same process for myself, hence asking for some help here.

So I have created 3 script files:
1. bg.py --> here I just keep a background image which I want to display across all pages, hence keeping it separately here
2. login.py --> this one is to show a login card, with username, password etc
3. app.py --> this is just the main file to run the app

So here is how I am putting the codes in each of the files:

**bg.py**
```python
pic =56
image = f'https://picsum.photos/id/{pic}/1900/920'

def bg():
    ui.image(image).classes('rounded-borders')
```

**login.py**
```python
def login_card():
    with ui.card().classes('mt-40 ml-40'):  
        ui.input(placeholder="Username")
        ui.input(placeholder='Password', password=True, password_toggle_button=True)
```

Then I try to call them together using the **app.py**
```python
@ui.page('/')
def home_page():
    with bg.bg():
        login.login_card()

home_page()
```

But this does not work. I saw in the **Modularization** example, there is the ```@contextmanager``` decorator used, so I tried that one also, but did not fully understand how to use it. Is the error happening because of that?

Kindly let me know the right way of doing this please. Basically what I am trying to achieve is, I create all the different page elements like the background, sidebars etc. in separate modules, and then import them as needed to create the pages, so that later on I don't need to make changes in multiple places.

Thanks,
Anindya

**Comment by Anindya088** [comment]
Sorry I think I missed out giving some more details, and did not explain what the issue is happening. Apologies, so the main issue is that the login card is not overlaying on top of the image, it's coming below it as the next element. Sharing screenshot for reference:
![image](https://github.com/zauberzeug/nicegui/assets/65127974/bbc48e13-bf0b-4e39-b139-e311bcc77cd0)

I tried using the ```contextmanager``` in this way in the **bg.py**
```python
from contextlib import contextmanager
pic =56
image = f'https://picsum.photos/id/{pic}/1900/920'

@contextmanager
def bg():
    ui.image(image).classes('rounded-borders')
    yield
```

Please let me know what I am doing wrong, and how to overlay the login card on top of the image.

**Comment by rodja** [comment, marked as answer]
Your `bg()` function should return the image element. That way you can call the function and use it as a context for the login:

```py
def bg():
    return ui.image('https://picsum.photos/id/377/640/360').classes('rounded-borders')


def login_card():
    with ui.card().classes('mt-40 ml-40'):
        ui.input(placeholder="Username")
        ui.input(placeholder='Password', password=True, password_toggle_button=True)


@ui.page('/')
def home_page():
    with bg():
        login_card()
```

By the way: you do not need to call your page generator with `home_page()`. With the `@ui.page` decorator the function will be registered in FastAPI to be executed for each user who accesses the provided route (eg. `/`).

**Reply by Anindya088** [reply]
Hi @rodja , thanks a lot. I was thinking, how to do this when I try to keep the codes in separate files and import the modules?

Like let's say the function bg() and login_card() are in 2 different files. Then maybe there is a third file, let's say app.py where I import the other 2 files and call the functions. I realized the "with" style is not working then, most likely I'm doing something wrong there. Because the code runs but the login card comes below the image.

Can you please advise on how to do this correctly if the image and login card functions are imported from other files?

Thanks a lot,
Anindya

**Reply by rodja** [reply]
I just packed it all in one file for better try out. The mechanism is the same if you have separate files. Just return the image in your `bg()` function within the `bg.py` file.

**Reply by Anindya088** [reply]
Hi @rodja , then I must be doing something wrong, but can't seem to find out what the issue is.

If you look at my original screenshot with the code examples, I have split them into 3 files and tried calling them from the main file app.py but the login card seems to come below the image.

I tried the context manager as well, the code example in the first reply. But does not seem to work.

Any idea what I'm missing or doing wrong?

Thanks,
Anindya

**Reply by rodja** [reply]
@Anindya088 Have you tried my suggestion? As I said, with the return statement it works also with multiple files. Please compare this example with yours: [nest_elements.zip](https://github.com/zauberzeug/nicegui/files/11780640/nest_elements.zip)

**Reply by Anindya088** [reply]
HI @rodja , sorry sorry, my apologies, I completely missed that I am not using the ```return``` statement.
So sorry, in fact I did not even check my code properly until you prompted again, I was thinking I had it.

It's working perfectly fine with the ```return``` statement. Sorry for wasting your Sunday on this!

Thanks a lot,
Anindya
